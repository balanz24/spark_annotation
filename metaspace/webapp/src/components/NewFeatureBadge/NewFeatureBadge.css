.sm-feature-badge sup {
  opacity: 1;
  transition: 0.28s cubic-bezier(0.4, 0, 0.2, 1);
  transition-property: opacity;
}

.sm-feature-badge--hidden sup {
  opacity: 0;
}
