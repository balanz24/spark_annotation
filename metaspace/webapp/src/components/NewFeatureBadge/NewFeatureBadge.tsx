import './NewFeatureBadge.css'

import Vue from 'vue'
import { defineComponent, reactive } from '@vue/composition-api'

import { getLocalStorage, setLocalStorage } from '../../lib/localStorage'

const storageKey = 'new_feature_badges'

const store: { [key: string]: boolean } = reactive(
  getLocalStorage(storageKey) || {},
)

export function hideFeatureBadge(featureKey: string) {
  if (store[featureKey]) {
    return
  }
  Vue.set(store, featureKey, true)
  setLocalStorage(storageKey, store)
}

const NewFeatureBadge = defineComponent({
  name: 'NewFeatureBadge',
  props: {
    featureKey: { type: String, required: true },
    showUntil: Date as any as () => Date,
  },
  setup(props, { slots }) {
    const isStale = props.showUntil && props.showUntil.valueOf() < Date.now()
    if (!(props.featureKey in store)) {
      Vue.set(store, props.featureKey, false)
    }
    if (store[props.featureKey] || isStale) {
      return () => slots.default()
    }
    return () => (
      <el-badge
        value="New"
        class={['sm-feature-badge', { 'sm-feature-badge--hidden': store[props.featureKey] }]}
      >
        {slots.default()}
      </el-badge>
    )
  },
})

export default NewFeatureBadge
