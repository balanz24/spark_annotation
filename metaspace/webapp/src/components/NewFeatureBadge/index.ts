import './NewFeatureBadge.css'

export * from './NewFeatureBadge'
export { default } from './NewFeatureBadge'
