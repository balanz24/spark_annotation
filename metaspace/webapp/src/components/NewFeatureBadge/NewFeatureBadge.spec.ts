import { mount } from '@vue/test-utils'
import Vue from 'vue'
import NewFeatureBadge, { hideFeatureBadge } from './NewFeatureBadge'

const TestNewFeatureBadge = Vue.component('test', {
  functional: true,
  render: (h, { props }) => h(NewFeatureBadge, { props }, [h('span', {}, 'Badge Content')]),
})

const featureKey = 'test'

describe('NewFeatureBadge', () => {
  it('should render a badge', () => {
    const wrapper = mount(TestNewFeatureBadge, { propsData: { featureKey } })
    expect(wrapper.classes()).toMatchSnapshot()
  })

  it('should hide the badge', async() => {
    const wrapper = mount(TestNewFeatureBadge, { propsData: { featureKey } })
    hideFeatureBadge(featureKey)
    await Vue.nextTick()
    expect(wrapper.classes()).toMatchSnapshot()
  })

  it('should not render the badge if already hidden', () => {
    hideFeatureBadge(featureKey)
    const wrapper = mount(TestNewFeatureBadge, { propsData: { featureKey } })
    expect(wrapper.classes()).toMatchSnapshot()
  })

  it('should not render the badge if stale', () => {
    const spy = jest.spyOn(global.Date, 'now').mockImplementation(() => new Date('2020-01-02T00:00:00.000').valueOf())
    const wrapper = mount(TestNewFeatureBadge, {
      propsData: { featureKey, showUntil: new Date('2020-01-01T00:00:00.000') },
    })
    expect(wrapper.classes()).toMatchSnapshot()
    spy.mockRestore()
  })
})
