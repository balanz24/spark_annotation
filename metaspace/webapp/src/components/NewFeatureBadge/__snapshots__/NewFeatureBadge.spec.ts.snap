// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewFeatureBadge should hide the badge 1`] = `
Array [
  "el-badge",
  "sm-feature-badge",
  "sm-feature-badge--hidden",
]
`;

exports[`NewFeatureBadge should not render the badge if already hidden 1`] = `Array []`;

exports[`NewFeatureBadge should not render the badge if stale 1`] = `Array []`;

exports[`NewFeatureBadge should render a badge 1`] = `
Array [
  "el-badge",
  "sm-feature-badge",
]
`;
